packages/jest-runner/src/runTest.ts

```diff
@@ -109,9 +108,26 @@ async function runTestInternal(
 
   const cacheFS = new Map([[path, testSource]]);
   const transformer = await createScriptTransformer(projectConfig, cacheFS);
+  const testEnvironmentIsEsm = Resolver.unstable_shouldLoadAsEsm(
```

Comment:
The test `e2e/__tests__/testEnvironmentEsm.ts` still failed because this line activates `cachedShouldLoadAsEsm` which somehow doesn't accept new value of `projectConfig.extensionsToTreatAsEsm`. 

Can you debug and investigate why? If we can't change here, where should we change?